pkg/logs/input/docker/tailer.go

```diff
@@ -127,15 +128,30 @@ func (t *Tailer) tail(since string) error {
 // readForever reads from the reader as fast as it can,
 // and sleeps when there is nothing to read
 func (t *Tailer) readForever() {
+	c := make(chan struct{}, 1)
 	defer t.decoder.Stop()
 	for {
+		var n int
+		var err error
+
 		select {
 		case <-t.stop:
 			// stop reading new logs from container
 			return
 		default:
 			inBuf := make([]byte, 4096)
-			n, err := t.reader.Read(inBuf)
+			// Read can be blocking because it's a wrapper over an HTTP call
+			// If the Read() timeouts, the tailer will be restarted.
+			go func() {
+				n, err = t.reader.Read(inBuf)
```

Comment:
I am surprised that a reader does not support a timeout natively, did you check already ?

Reply:
Yes, the reader implement the `io.ReadCloser` interface that only has `Read()` and `Close()` methods

The only way to do it would be to create a new Reader each time we read, with a context with a timeout (cf the `setupReader` method)

It adds overheads because every `timeout` duration you have to `Close()` the reader and open another, regardless of the reader status (blocked or not)

It requires that we also keep track of the current offset in the tailer